### ISSUE TITLE
Source maps for wasm2c?

### ISSUE THREAD

**Issue by SoniEx2** [WebAssembly/wabt#2215]
Does it... make any sense to have source maps for wasm2c?

**Comment by keithw** [comment]
It would be nice to have _something_ like this, because it's pretty challenging to debug a program that's been compiled from, e.g., C++ (via g++) to Wasm (and then, via wasm2c) to C (and then, via gcc) to ELF. See #1210 & #432.

An MVP might be to make wasm2c include little comments next to each line of a function body, giving a Wat version of the Wasm instruction, so that if you then compile wasm2c's output with `-g`, a debugger that speaks DWARF can show the Wat for each line. That seems like a one-day project.

But it would be really nice to plumb it all the way through somehow so that a debugger of the final object file can show the original high-level source code.

**Comment by sbc100** [comment]
I think we could use C pre-processor directives: https://gcc.gnu.org/onlinedocs/cpp/Line-Control.html

**Comment by shravanrn** [comment]
yup, the C-preprocessor debug directives was something I was hoping to do for several months now, but I don't think I have time on my roadmap. I would love to see support for something like this too!

**Comment by turbolent** [comment]
In https://github.com/turbolent/w2c2 we parse the DWARF debug info (if provided) from the module, and generate C `#line` directives. It works fairly well

**Comment by shravanrn** [comment]
(Sorry for the delay in responding was traveling)
Ah this is good to know. I will add it to my todo list to see if we can get something like this as well.
Assuming the licenses are compatible, and there are no objections, I am hoping we can borrow some of the implementation from w2c2 for this effort.

**Comment by squk** [comment]
This would be a really useful feature! Deploying sandboxed libraries to 100% in a large production environment without proper symbolization is going to be challenging. Especially when the native code is under development by a separate contributing team. For example, source maps would allow code paths in an "embedder"'s middleware(e.g. additional WASI impls or other WASM module imports used in wasm2c code) to be differentiated from code paths in the contributor teams code being sandboxed.
